src/backend/distributed/metadata/pg_get_object_address_12_13_14.c

```diff
@@ -0,0 +1,528 @@
+
+			owa->objname = name;
+			owa->objargs = args;
+			objnode = (Node *) owa;
+			break;
+		}
+
+		case OBJECT_LARGEOBJECT:
+		{
+			/* already handled above */
+			break;
+		}
+
+			/* no default, to let compiler warn about missing case */
+	}
+
+	if (objnode == NULL)
+	{
+		elog(ERROR, "unrecognized object type: %d", type);
+	}
+
+	ObjectAddress addr = get_object_address(type, objnode,
+											&relation, AccessShareLock, false);
+
+	/* CITUS CODE BEGIN */
+	ErrorIfCurrentUserCanNotDistributeObject(type, &addr, objnode, &relation);
+
+	/* CITUS CODE END */
+
+	/* We don't need the relcache entry, thank you very much */
+	if (relation)
+	{
+		relation_close(relation, AccessShareLock);
+	}
+
+	/* CITUS CODE BEGIN */
+	return addr;
+
+	/* CITUS CODE END */
+}
+
+
+/*
+ * ErrorIfCurrentUserCanNotDistributeObject checks whether current user can
+ * distribute object, if not errors out.
+ */
+static void
+ErrorIfCurrentUserCanNotDistributeObject(ObjectType type, ObjectAddress *addr,
+										 Node *node, Relation *relation)
+{
+	Oid userId = GetUserId();
+	AclMode aclMaskResult = 0;
+	bool skipAclCheck = false;
+	Oid idToCheck = InvalidOid;
+
+	/* Since we don't handle sequences like object, add it separately */
+	if (!(SupportedDependencyByCitus(addr) || type == OBJECT_SEQUENCE))
+	{
+		ereport(ERROR, (errmsg("Object type %d can not be distributed by Citus", type)));
+	}
+
+	switch (type)
+	{
+		case OBJECT_SCHEMA:
+		{
+			idToCheck = addr->objectId;
+			aclMaskResult = pg_namespace_aclmask(idToCheck, userId, ACL_USAGE,
+												 ACLMASK_ANY);
+			break;
+		}
+
+		case OBJECT_FUNCTION:
+		case OBJECT_PROCEDURE:
+		case OBJECT_AGGREGATE:
+		{
+			check_object_ownership(userId, type, *addr, node, *relation);
+			skipAclCheck = true;
+			break;
+		}
+
+		case OBJECT_DATABASE:
+		{
+			idToCheck = addr->objectId;
+			aclMaskResult = pg_database_aclmask(idToCheck, userId, ACL_CONNECT,
+												ACLMASK_ANY);
+			break;
+		}
+
+		case OBJECT_ROLE:
+		{
+			/* Support only extension owner role with community */
+			if (addr->objectId != CitusExtensionOwner())
+			{
+				ereport(ERROR, (errmsg("Current user does not have required "
+									   "access privileges on role %d with type %d",
+									   addr->objectId, type)));
+			}
+			skipAclCheck = true;
+			break;
+		}
+
+		case OBJECT_TYPE:
+		{
+			idToCheck = addr->objectId;
+			aclMaskResult = pg_type_aclmask(idToCheck, userId, ACL_USAGE,
+											ACLMASK_ANY);
+			break;
+		}
+
+		case OBJECT_SEQUENCE:
+		{
+			idToCheck = addr->objectId;
+			aclMaskResult = pg_class_aclmask(idToCheck, userId, ACL_USAGE, ACLMASK_ANY);
+			break;
+		}
+
+		case OBJECT_TABLE:
+		{
+			/* table distribution already does the ownership check, so we can stick to that over acl_check */
+			check_object_ownership(userId, type, *addr, node, *relation);
+			skipAclCheck = true;
+			break;
+		}
+
+		case OBJECT_EXTENSION:
+		{
+			/* as of PG 14, only super users can create extension */
```

Comment:
Ops, this seems wrong. With PG13 +, some extensions can be created by regular users with `CREATE` grant. See [docs](https://www.postgresql.org/docs/14/sql-createextension.html):
```
Loading an extension ordinarily requires the same privileges that would be required to create its component objects. For many extensions this means superuser privileges are needed. However, if the extension is marked trusted in its control file, then it can be installed by any user who has CREATE privilege on the current database. In this case the extension object itself will be owned by the calling user, but the contained objects will be owned by the bootstrap superuser (unless the extension's script explicitly assigns them to the calling user). This configuration gives the calling user the right to drop the extension, but not to modify individual objects within it.
```
So, contrib extensions (e.g., `trusted extensions`) can be created with database create usage acls. We should find the same checks that PG does.

```SQL
CREATE ROLE my_role NOSUPERUSER CREATEDB CREATEROLE INHERIT LOGIN;
GRANT CREATE ON DATABASE postgres TO my_role ;
SELECT 1 FROM run_command_on_workers($$CREATE ROLE my_role NOSUPERUSER CREATEDB CREATEROLE INHERIT LOGIN;$$);
SELECT 1 FROM run_command_on_workers($$GRANT CREATE ON DATABASE postgres TO my_role ;;$$);
\c - my_role - 
You are now connected to database "postgres" as user "my_role".
set citus.log_remote_commands to on;
CREATE EXTENSION isn ;
```